Kernel/MismatchHandler.hpp

```diff
@@ -18,48 +18,107 @@
 
 #include "Forwards.hpp"
 #include "Term.hpp"
+#include "Kernel/TermTransformer.hpp"
+#include "Lib/MaybeBool.hpp"
+#include "Lib/BiMap.hpp"
 
 namespace Kernel
 {
 
-class MismatchHandler
+class MismatchHandler : public TermTransformer
 {
 public:
+
+  MismatchHandler() : TermTransformer(false) {}
+
   // returns true if the mismatch was handled.
-  virtual bool handle(RobSubstitution* sub, TermList t1, unsigned index1, TermList t2, unsigned index2) = 0;
+  virtual bool handle(TermList t1, unsigned index1, TermList t2, unsigned index2, 
+    UnificationConstraintStack& ucs, BacktrackData& bd, bool recording) = 0;
+  virtual TermList transformSubterm(TermList t) = 0;
+
+  // With polymorphism, a term may end up being a constraint term
+  // depending on type substitutions.
+  // Also a term such as f(a,b) : $int may be a constraint term
+  //
+  // Function used as follows:
+  // if returns true, term is ONLY involved in constraints and never unified
+  // if returns false, term is ONLY unified and never involved in constraints
+  // if returns maybe, term is involved in constraints and unified
+  virtual MaybeBool isConstraintTerm(TermList t) = 0;
+
+  bool areIdentical(Term* t1, Term* t2, unsigned idx1, unsigned idx2);
+
+  virtual Term* get(unsigned var){ NOT_IMPLEMENTED; }
+
```

Comment:
We do use `get` in release code, but only via the `compositeHandler` since no class should have direct access to any handler object other than `compositeHandler`.

For unit testing, we take shortcuts and don;t bother creating a `compositeHandler` which is why other handlers need to implement `get`